gax/operation.h

```diff
@@ -0,0 +1,118 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GAPIC_GENERATOR_CPP_GAX_OPERATION_H_
+#define GAPIC_GENERATOR_CPP_GAX_OPERATION_H_
+
+#include "call_context.h"
+#include "internal/gtest_prod.h"
+#include "operations_stub.h"
+#include "status.h"
+#include "status_or.h"
+
+#include "google/longrunning/operations.pb.h"
+
+#include <iostream>
+
+#include <memory>
+
+namespace google {
+
+namespace gax {
+
+// Note: compilation will fail unless ResponseT and MetadataT are children of
+// protobuf::Message.
+template <typename ResponseT, typename MetadataT>
+class Operation final {
+ public:
+  Operation(std::shared_ptr<gax::OperationsStub> stub,
+            google::longrunning::Operation op)
+      : stub_(std::move(stub)), op_(std::move(op)) {}
+  Operation() = delete;
+  Operation(Operation&& rhs) = default;
+
+  std::string const& Name() { return op_.name(); }
+
+  gax::StatusOr<ResponseT> Result() {
+    if (!Done()) {
+      return gax::Status{gax::StatusCode::kUnknown,
+                         "operation has not completed=" + Name()};
+    } else if (op_.has_error()) {
+      return gax::Status{static_cast<gax::StatusCode>(op_.error().code()),
+                         op_.error().message()};
+    } else {
+      auto const& any = op_.response();
+      if (!any.Is<ResponseT>()) {
+        return gax::Status{gax::StatusCode::kUnknown,
+                           "invalid result in operation=" + Name()};
+      }
+
+      ResponseT result;
+      any.UnpackTo(&result);
+      return std::move(result);
+    }
+  }
+
+  MetadataT Metadata() const {
+    MetadataT m;
+    op_.metadata().UnpackTo(&m);
+    return m;
+  }
+
+  gax::Status Update() {
+    if (!Done()) {
+      google::longrunning::GetOperationRequest request;
+      request.set_name(op_.name());
+      gax::CallContext context(OperationsStub::get_operation_info_);
+      return stub_->GetOperation(context, request, &op_);
+    } else if (op_.has_error()) {
+      return gax::Status{static_cast<gax::StatusCode>(op_.error().code()),
+                         op_.error().message()};
+    } else {
+      return gax::Status{};
+    }
+  }
+
+  gax::Status Delete() {
+    google::longrunning::DeleteOperationRequest request;
+    google::protobuf::Empty empty;
+    request.set_name(op_.name());
+    gax::CallContext context(OperationsStub::delete_operation_info_);
+    return stub_->DeleteOperation(context, request, &empty);
+  }
+
+  gax::Status Cancel() {
+    google::longrunning::CancelOperationRequest request;
+    google::protobuf::Empty empty;
+    request.set_name(op_.name());
+    gax::CallContext context(OperationsStub::cancel_operation_info_);
+    return stub_->CancelOperation(context, request, &empty);
+  }
+
+  bool Done() { return op_.done(); }
+
+ private:
+  FRIEND_TEST(Operation, Basic);
```

Comment:
Why do you need the `FRIEND_TEST`?  Is it possible to write the tests without these friend declarations?  Also, does this mean I need to install googletest to even *use* your library?  What if I prefer to use `Catch` as my test infrastructure?

Reply:
At the moment, gax provides its own definition of the header and `FRIEND_TEST` macro because gtest-1.8.1 doesn't provide fine grained targets, so there is no extra dependency for users. This is subject to change, however.
I can add additional accessor methods to remove the need for `FREIND_TEST`, but adjusting the public interface to facilitate testing feels a little bit wrong. Alternatively, manipulating `op_` via the methods of `DummyOperationsStub` removes the `FRIEND_TEST` dependency, but writing the tests becomes much harder.

Both gax and the generated client are required to be exception-free, so `Catch` isn't available.

Reply:
The alternative is to write your tests such they do not require examining the internal state of the class, right?  Is that possible?

Reply:
It's possible, byt the tricky part is that I want to _modify_ internal state. The 'correct' way to do this would be to make a new custom OperationsStub child class per test whose `GetOperation` method takes the g::l::Operation object by ptr and modifies it in the correct way.

If you feel strongly enough about removing `FRIEND_TEST` I'll absolutely do this.

Reply:
It is more of a guideline: go/unit-test-practices#public-apis